### PR TITLE
[SPARK-39678][SQL] Improve stats estimation for v2 tables

### PR DESCRIPTION
**Pull request by singhpk234**
### What changes were proposed in this pull request?

We should propagate the row count stats in SizeInBytesOnlyStatsPlanVisitor if available. Row counts are propagated from connectors to spark in case of v2 tables.

### Why are the changes needed?

This can improve stats estimation for v2 tables, since row count is used at places to estimate sizeInBytes.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Modified existing UT's to match the proposed behavior.

### CODE REVIEW

**Comment by singhpk234** [comment]
cc @huaxingao @cloud-fan @wangyum

**Review by wangyum — COMMENTED** [review]
Could you enable `spark.sql.cbo.enabled` to estimate row count?
https://github.com/apache/spark/blob/b1d719e7c9faeb5661a7e712b3ecefca56bf356f/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/LogicalPlanStats.scala#L33-L40

**Comment by singhpk234** [comment]
> Could you enable spark.sql.cbo.enabled to estimate row count?

Thanks @wangyum, I am aware of the alternate visitor we use with cbo. 

I raised this pr considering : 
1. cbo is turned off by default.
2. We already have rowCount propagated via LeafNodes (DSv2Relation) which are used for estimating output size in SizeInBytesOnlyStatsPlanVisitor
https://github.com/apache/spark/blob/161c596cafea9c235b5c918d8999c085401d73a9/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/SizeInBytesOnlyStatsPlanVisitor.scala#L93-L100
3. ANALYZE is not supported for v2 tables so except row count, IMHO we can't have ndv etc. I am refering to this jira : https://issues.apache.org/jira/browse/SPARK-39420
4. As per my understanding v1 tables can only pass in sizeInBytes unless they have some stats in catalog. whereas v2 tables already give both from the relation itself, hence I thought it's un-accounted for v2 tables.
https://github.com/apache/spark/blob/161c596cafea9c235b5c918d8999c085401d73a9/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/LogicalRelation.scala#L43-L45


Are you recommending it's an expected behavior / by design ?

**Comment by singhpk234** [comment]
rebased and regenerated the golden files via : 
 *   SPARK_GENERATE_GOLDEN_FILES=1 build/sbt "sql/testOnly *PlanStability*Suite"
 *   SPARK_GENERATE_GOLDEN_FILES=1 SPARK_ANSI_SQL_MODE=true build/sbt "sql/testOnly *PlanStability*Suite"

**Comment by wangyum** [comment]
I think it's by design. So enabling `spark.sql.cbo.enabled` is what you want?

**Comment by singhpk234** [comment]
Thanks @wangyum !

> So enabling spark.sql.cbo.enabled is what you want?

I believe then setting `spark.sql.cbo.enabled` to true by default could help, (what i wanted was to take this valuable stat of row count, bubbled up from v2 connector to be accounted for in default spark behaviour) but I think it requires some additional efforts, since our other defaults such as auto-bhj etc needs to adjusted accordingly.

> I think it's by design

for my knowledge, can you please point me to some jira's ,happy to learn more.

Love to know your thoughts on the same, Happy to close this as well if we consider this is not a problem at all.

**Comment by cloud-fan** [comment]
I'm a bit confused. After this PR, what's the difference between `SizeInBytesOnlyStatsPlanVisitor` and `BasicStatsPlanVisitor`?

**Comment by singhpk234** [comment]
> After this PR, what's the difference between SizeInBytesOnlyStatsPlanVisitor and BasicStatsPlanVisitor

BasicStatsPlanVisitor additionally takes has columnStats such as (NDV / NullCount / min / max etc) on estimation, which generally is not passed from DSv1 / Dsv2 relation itself.  

As per my understanding, prior to this PR, SizeInBytesOnlyStatsPlanVisitor was estimating stats on the subset of info i.e only sizeInBytes and BasicStatsPlanVisitor on all 3 info (sizeInBytes, rowcount,ColumStats (min /max /NDV etc), now via this PR SizeInBytesOnlyStatsPlanVisitor is estimating stats on the subset of info but this subset is now (sizeInBytes / rowCount) and BasicStatsPlanVisitor on all 3 info (sizeInBytes, rowcount,ColumStats (min /max /NDV etc).

**Comment by cloud-fan** [comment]
Maybe we should name them `BasicStatesPlanVisitor` and `BasicAndColumnStatsPlanVisitor`. We also need to make sure the updated `SizeInBytesOnlyStatsPlanVisitor` can propagate row count properly in all cases.

BTW, with CBO off, where do we use row count?

**Comment by singhpk234** [comment]
> BTW, with CBO off, where do we use row count?

we use it in places like : https://github.com/apache/spark/blob/161c596cafea9c235b5c918d8999c085401d73a9/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/SizeInBytesOnlyStatsPlanVisitor.scala#L93-L100

where we just multiply row-count with row size. In v1 scenario in case of [logical relation row-count](https://github.com/apache/spark/blob/161c596cafea9c235b5c918d8999c085401d73a9/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/LogicalRelation.scala#L43-L45) can seep in from catalog stats but it has a has a chance of `row-count` getting lost in places where we assume we only have sizeInBytes for example here : 
https://github.com/apache/spark/blob/161c596cafea9c235b5c918d8999c085401d73a9/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/SizeInBytesOnlyStatsPlanVisitor.scala#L54-L58

**Comment by cloud-fan** [comment]
OK I think the idea makes sense. With CBO off, the optimizer/planner only needs size in bytes, but row count is also an important statistics to estimate size in bytes, and should be propagated in the stats plan visitor.

**Comment by cloud-fan** [comment]
cc @wzhfy @c21 can you take a look first?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/AdvancedStatsPlanVisitor.scala [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+  override def visitGenerate(p: Generate): Statistics = default(p)
+
+  override def visitGlobalLimit(p: GlobalLimit): Statistics = fallback(p)
+
+  override def visitOffset(p: Offset): Statistics = fallback(p)
+
+  override def visitIntersect(p: Intersect): Statistics = fallback(p)
+
+  override def visitJoin(p: Join): Statistics = fallback(p)
```
Comment (c21) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/AdvancedStatsPlanVisitor.scala:57]: Why we fallback here, but not use `JoinEstimation.estimate`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/AdvancedStatsPlanVisitor.scala [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.plans.logical.statsEstimation
+
+import org.apache.spark.sql.catalyst.plans.logical._
+
+/**
+ * A [[LogicalPlanVisitor]] that computes the statistics for the cost-based optimizer.
```
Comment (c21) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/AdvancedStatsPlanVisitor.scala:23]: Can you also mention here, what are the statistics we compute here? e.g. `sizeInBytes`, `rowCount` etc, similar to the top-level comment of `BasicStatsPlanVisitor`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/BasicStatsPlanVisitor.scala [hunk 3]
```diff
@@ -17,16 +17,40 @@
 
 package org.apache.spark.sql.catalyst.plans.logical.statsEstimation
 
+import org.apache.spark.sql.catalyst.expressions.AttributeMap
 import org.apache.spark.sql.catalyst.plans.logical._
 
 /**
- * A [[LogicalPlanVisitor]] that computes the statistics for the cost-based optimizer.
+ * An [[LogicalPlanVisitor]] that computes a single dimension for plan stats: size in bytes.
```
Comment (c21) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/BasicStatsPlanVisitor.scala:24]: nit: typo `An` -> `A`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/BasicStatsPlanVisitor.scala [hunk 4]
```diff
@@ -17,16 +17,40 @@
...
+    var sizeInBytes = (p.child.stats.sizeInBytes * outputRowSize) / childRowSize
+    if (sizeInBytes == 0) {
+      // sizeInBytes can't be zero, or sizeInBytes of BinaryNode will also be zero
+      // (product of children).
+      sizeInBytes = 1
+    }
+
+    // v2 sources can bubble-up rowCount, so always propagate.
+    // Don't propagate attributeStats, since they are not estimated here.
+    Statistics(sizeInBytes = sizeInBytes, rowCount = p.child.stats.rowCount)
```
Comment (c21) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/BasicStatsPlanVisitor.scala:47]: I am confused here. In the top-level comment - `computes a single dimension for plan stats: size in bytes`. But why we populate `rowCount` as well here?

**Review by c21 — COMMENTED** [review]
Thanks [singhpk234](https://github.com/singhpk234) for the work! Having some comments/questions.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/BasicStatsPlanVisitor.scala (hunk 4 above)
Comment (singhpk234) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/BasicStatsPlanVisitor.scala:47]: In this estimator, we adjust the size by scaling it by (input row size / output row size) but since we don't have much info (in terms of min / max / ndv etc) to estimate the row count we just say the node output's child output row count which is mostly true for operators like project etc.

Since we were just computing sizeInBytes and propagating rowCounts as it is I left the comment as it is.

Should I rephrase it to let's say : 
- `estimates size in bytes, row count for plan stats`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/AdvancedStatsPlanVisitor.scala (hunk 1 above)
Comment (singhpk234) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/AdvancedStatsPlanVisitor.scala:57]: fallback here would endup calling BasicStatsPlanVisitor.visit(p) which, will in turn call BasicStatsPlanVisitor#visitJoin which will be `JoinEstimation(p).estimate.getOrElse(default(p))`.  Hence added the same.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/BasicStatsPlanVisitor.scala (hunk 4 above)
Comment (singhpk234) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/BasicStatsPlanVisitor.scala:47]: for dsv2 sources rowCount can be passed from the relation itself without running `analyze`, hence BasicStatsPlanVisitor which will be our default now will take rowcount into consideration.

File: sql/core/src/test/resources/tpcds-plan-stability/approved-plans-v1_4/q44.sf100/explain.txt [hunk 5]
```diff
@@ -1,35 +1,41 @@
...
-      :     :                 +- * Sort (15)
-      :     :                    +- ReusedExchange (14)
-      :     +- BroadcastExchange (25)
-      :        +- * Filter (24)
-      :           +- * ColumnarToRow (23)
-      :              +- Scan parquet spark_catalog.default.item (22)
-      +- ReusedExchange (28)
+TakeOrderedAndProject (37)
++- * Project (36)
+   +- * SortMergeJoin Inner (35)
```
Comment (zinking) [inline sql/core/src/test/resources/tpcds-plan-stability/approved-plans-v1_4/q44.sf100/explain.txt:4]: is this expected ? looks like a plan regression to me

**Comment by zinking** [comment]
> > BTW, with CBO off, where do we use row count?
> 
> we use it in places like :
> 
> https://github.com/apache/spark/blob/161c596cafea9c235b5c918d8999c085401d73a9/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/SizeInBytesOnlyStatsPlanVisitor.scala#L93-L100
> 
> where we just multiply row-count with row size. We also use it for BF to create [bloomFilterAgg](https://github.com/apache/spark/blob/13882bd7b80cd89fc4c58bd96a5ef783a0744019/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/InjectRuntimeFilter.scala#L78-L84). In v1 scenario in case of [logical relation row-count](https://github.com/apache/spark/blob/161c596cafea9c235b5c918d8999c085401d73a9/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/LogicalRelation.scala#L43-L45) can seep in from catalog stats but as you correctly pointed out it has a has a chance of `row-count` getting lost in places where we assume we only have sizeInBytes for example here :
> 
> https://github.com/apache/spark/blob/161c596cafea9c235b5c918d8999c085401d73a9/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/SizeInBytesOnlyStatsPlanVisitor.scala#L54-L58

thought these stats are available in AQE and more accurate though
